muffnn/autoencoder/autoencoder.py

```diff
@@ -0,0 +1,686 @@
+            start_idx = 0
+            epoch = 0
+            indices = np.arange(n_examples)
+            random_state.shuffle(indices)
+
+            while True:
+                batch_ind = indices[start_idx:start_idx + self.batch_size]
+                feed_dict = self._make_feed_dict(X[batch_ind], training=True)
+                obj_val, _ = self._session.run(
+                    [self._obj_func, self._train_step], feed_dict=feed_dict)
+                _LOGGER.info("objective: %.4f, epoch: %d, idx: %d",
+                             obj_val, epoch, start_idx)
+                start_idx += self.batch_size
+                if start_idx > n_examples - self.batch_size:
+                    start_idx = 0
+                    epoch += 1
+                    if epoch >= self.n_epochs:
+                        break
+                    random_state.shuffle(indices)
+
+        return self
+
+    def transform(self, X, y=None):
+        """Encode data with the autoencoder.
+
+        Parameters
+        ----------
+        X : numpy array or sparse matrix of shape [n_samples, n_features]
+            Data to encode
+
+        Returns
+        -------
+        Z : numpy array
+            Encoded data.
+        """
+
+        if not self._is_fitted:
+            raise NotFittedError("Call fit before transform!")
+
+        # For sparse input, make the input a CSR matrix since it can be
+        # indexed by row.
+        X = check_array(X, accept_sparse=['csr'])
+
+        # Check input data against internal data.
+        # Raises an error on failure.
+        self._check_data(X)
+
+        # Make predictions in batches.
+        pred_batches = []
+        start_idx = 0
+        n_examples = X.shape[0]
+        with self.graph_.as_default():
+            while start_idx < n_examples:
+                X_batch = \
+                    X[start_idx:min(start_idx + self.batch_size, n_examples)]
+                feed_dict = self._make_feed_dict(X_batch, training=False)
+                start_idx += self.batch_size
+                pred_batches.append(self._session.run(self._encoded_values,
+                                                      feed_dict=feed_dict))
+        return np.concatenate(pred_batches)
+
+    def inverse_transform(self, X, y=None, return_sparse=False):
+        """Decode data with the autoencoder.
+
+        Note that transform and inverse_transform will in general not be exact
+        inverses of each other for an autoencoder.
+
+        This method can be useful if further modeling on the encoded data
+        generates new encoded data vectors (e.g., generating new data by
+        fitting a density estimate to the encoded data).
+
+        Parameters
+        ----------
+        X : numpy array or sparse matrix of shape [n_samples, n_features]
+            Encoded data to decode
+
+        return_sparse : bool, optional
+            If True, return a sparse matrix.
+
+        Returns
+        -------
+        Z : numpy array
```

Comment:
To indicate that this is encoded data. I have seen this notation and using X would be confusing since that is an input.